larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -326,19 +373,781 @@ StatusCode MipEnergyFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+            }
+            featureVector.push_back(pca1);
+            featureVector.push_back(pca2);
+        }
+        else
+        {
+            featureVector.push_back(principalEigenvalue);
+            featureVector.push_back(secondaryEigenvalue);
+            featureVector.push_back(tertiaryEigenvalue);
+        }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode PCAFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+{
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "RatioVariables", m_ratioVariables));
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+ChargeFeatureTool::ChargeFeatureTool() :
+    m_ratioVariables(true),
+    m_addTotalCharge(false),
+    m_addChargeSigma(true),
+    m_addChargeIncrements(false),
+    m_addStartCharge(false),
+    m_addEndCharge(true),
+    m_endChargeFraction(0.1f)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ChargeFeatureTool::CalculateChargeVariables(const Algorithm *const pAlgorithm, const pandora::Cluster *const pCluster, float &totalCharge, float &chargeSigma, 
+    float &chargeMean, float &startCharge, float &endCharge, float &incrementCharge, float endChargeFraction)
+{
+
+    CaloHitList orderedCaloHitList(ChargeFeatureTool::OrderCaloHitsByDistanceToVertex(pAlgorithm, pCluster));
+
+    const int totalHits(pCluster->GetNCaloHits());
+    FloatVector chargeVector;
+    int hitCounter(0);
+    totalCharge = 0.f;
+    startCharge = 0.f;
+    endCharge = 0.f;
+    incrementCharge = 0.f;
+    float previousCharge(0.f);
+
+    for (const CaloHit *const pCaloHit : orderedCaloHitList)
+    {
+        hitCounter++;
+        const float pCaloHitCharge(pCaloHit->GetInputEnergy());
+
+        if (pCaloHitCharge < 0)
+        {
+            std::cout << "Found a hit with negative charge! " << std::endl;
+        }
+        else
+        {
+            totalCharge    += pCaloHitCharge;
+            chargeVector.push_back(pCaloHitCharge);
+
+            if (hitCounter <= std::ceil(totalHits*endChargeFraction)) //ensure at least 1 hit is used
+            {
+                startCharge += pCaloHitCharge;
+            }
+            if (hitCounter >= std::floor(totalHits*(1.f-endChargeFraction)))
+            {
+                endCharge += pCaloHitCharge;
+            }
+
+            if(hitCounter>1)
+                incrementCharge = (pCaloHitCharge-previousCharge)*(pCaloHitCharge-previousCharge);
```

Comment:
Why squared? Some kind of RMS-like variable?

Reply:
it is, but probably not worth keeping it, see previos message